File: merge_sort/Sorter.py
import sys

def merge_sort(arr: list, left: int, right: int) -> None:
    if left < right:
        mid = (left + right) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)

def merge(arr: list, left: int, mid: int, right: int) -> None:
    n1 = mid - left + 1
    n2 = right - mid

    L = []
    R = []

    for i in range(n1):
        L.append(arr[left + i])
    for i in range(n2):
        R.append(arr[mid + 1 + i])

    i = 0
    j = 0
    k = left
    while i < n1 and j < n2:
        if L[i] < R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1

    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1

def main():
    input_file = sys.argv[1] if len(sys.argv) > 1 else None
    output_file = sys.argv[2] if len(sys.argv) > 2 else None
    
    if input_file:
        f = open(input_file, 'r')            
        n_cases = int(f.readline())
    else:
        n_cases = int(sys.stdin.readline())
    
    results = []
    for _ in range(n_cases):
        line = f.readline() if input_file else sys.stdin.readline()
        arr = list(map(int, line.strip().split(" ")))
        merge_sort(arr, 0, len(arr) - 1)
        results.append(" ".join(map(str, arr)))
    
    if input_file: 
        f.close()

    if output_file:
        with open(output_file, 'w') as f:
            f.write("\n".join(results))
    else:
        print("\n".join(results))

if __name__ == "__main__":
    main()

File: merge_sort/Sorter.java
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.PrintStream;
import java.util.Arrays;
import java.util.Scanner;

public class Sorter {
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (right + left) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }
    
    private static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; i++) {
            L[i] = arr[left + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = arr[mid + 1 + j];
        }

        int i = 0;
        int j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k++] = L[i++];
            } else {
                arr[k++] = R[j++];
            }
        }
        while (i < n1) {
            arr[k++] = L[i++];
        }
        while (j < n2) {
            arr[k++] = R[j++];
        }
    }

    public static void main(String[] args) {
        Scanner scanner = null;
        if (args.length >= 1) {
            try {
                System.setIn(new FileInputStream(args[0]));
                scanner = new Scanner(System.in);
            } catch (FileNotFoundException e) {
                System.err.println("File not found " + args[0]);
            }
        } else {
            scanner = new Scanner(System.in);
        }
        if (args.length >= 2) {
            try {
                System.setOut(new PrintStream(args[1]));
            } catch (FileNotFoundException e) {
                System.err.println("File not found  " + args[1]);
                return;
            }
        }
        int cases = scanner.nextInt();
        String[] results = new String[cases];
        scanner.nextLine();
        for (int i = 0; i < cases; i++) {
            int[] arr = Arrays.stream(scanner.nextLine().trim().split(" "))
                              .mapToInt(Integer::parseInt)
                              .toArray();
            try{
                mergeSort(arr, 0, arr.length - 1);
                results[i] = String.join(" ", Arrays.stream(arr)
                                          .mapToObj(String::valueOf)
                                          .toArray(String[]::new));

            }
            catch(Exception e){
                results[i] = "Error" + e.getMessage();
            }
        }
        for (int i = 0; i < cases; i++) {
            System.out.println(results[i]);
        }
        scanner.close();
    }
}